src/backend/distributed/transaction/relation_access_tracking.c

```diff
@@ -0,0 +1,381 @@
+		{
+			continue;
+		}
+
+		RecordParallelSelectAccess(currentRelationId);
+
+		lastRelationId = currentRelationId;
+	}
+}
+
+
+/*
+ * RecordRelationMultiShardModifyAccessForTask gets a task and records
+ * the accesses. Note that the target relation is recorded with modify access
+ * where as the subqueries inside the modify query is recorded with select
+ * access.
+ */
+void
+RecordRelationMultiShardModifyAccessForTask(Task *task)
+{
+	List *relationShardList = NULL;
+	ListCell *relationShardCell = NULL;
+	Oid lastRelationId = InvalidOid;
+
+	/* no point in recoding accesses in non-transaction blocks, skip the loop */
+	if (!IsTransactionBlock())
+	{
+		return;
+	}
+
+	/* anchor shard is always associated with modify access */
+	RecordParallelModifyAccess(RelationIdForShard(task->anchorShardId));
+
+	if (task->modifyWithSubquery)
+	{
+		relationShardList = task->relationShardList;
+		foreach(relationShardCell, relationShardList)
+		{
+			RelationShard *relationShard = (RelationShard *) lfirst(relationShardCell);
+			Oid currentRelationId = relationShard->relationId;
+
+			/*
+			 * An optimization, skip going to hash table if we've already
+			 * recorded the relation.
+			 */
+			if (currentRelationId == lastRelationId)
+			{
+				continue;
+			}
+
+			RecordParallelSelectAccess(currentRelationId);
+
+			lastRelationId = currentRelationId;
+		}
+	}
+}
+
+
+/*
+ * RecordRelationMultiShardDDLAccessForTask marks all the relationShards
+ * with parallel DDL access if exists. That case is valid for inter-shard
+ * DDL commands such as foreign key creation. The function also records
+ * the relation that anchorShardId belongs to.
+ */
+void
+RecordRelationMultiShardDDLAccessForTask(Task *task)
+{
+	List *relationShardList = task->relationShardList;
+	ListCell *relationShardCell = NULL;
+
+	foreach(relationShardCell, relationShardList)
+	{
+		RelationShard *relationShard = (RelationShard *) lfirst(relationShardCell);
+
+		RecordParallelDDLAccess(relationShard->relationId);
+	}
+
+	RecordParallelDDLAccess(RelationIdForShard(task->anchorShardId));
+}
+
+
+/*
+ * RecordParallelSelectAccess is a wrapper around RecordParallelRelationAccess()
+ */
+void
+RecordParallelSelectAccess(Oid relationId)
+{
+	RecordParallelRelationAccess(relationId, PLACEMENT_ACCESS_SELECT);
+}
+
+
+/*
+ * RecordParallelModifyAccess is a wrapper around RecordParallelRelationAccess()
+ */
+void
+RecordParallelModifyAccess(Oid relationId)
+{
+	RecordParallelRelationAccess(relationId, PLACEMENT_ACCESS_DML);
+}
+
+
+/*
+ * RecordParallelDDLAccess is a wrapper around RecordParallelRelationAccess()
+ */
+void
+RecordParallelDDLAccess(Oid relationId)
+{
+	RecordParallelRelationAccess(relationId, PLACEMENT_ACCESS_DDL);
+}
+
+
+/*
+ * RecordParallelRelationAccess records the relation access mode as parallel
+ * for the given access type (e.g., select, dml or ddl) in the RelationAccessHash.
+ *
+ * The function becomes no-op for non-transaction blocks
```

Comment:
what about modifying CTEs?

Reply:
That's a very good point. I think in general I missed modifying `CTE`s in this implementation.

At first, I thought we could use `InCoordinatedTransaction()` instead of `IsTransactionBlock ()`, however, that also doesn't solve all of our problems, because if we do so, we'd skip the first `SELECT` statements which are router, because we don't call `BeginOrContinueCoordinatedTransaction()` for router selects.

```SQL
BEGIN;
SELECT count(*) FROM reference_Table;
DELETE from dist_table;
```

Also, using `InCoordinatedTransaction()` instead of `IsTransactionBlock()` would add some overhead for non-cte multi-shard queries, where we'd unnecessarly hit the hash tables.

Overall, seems like we should use `InCoordinatedTransaction() || IsTransactionBlock()`  to preserve the correctness, acknowledging the overhead. (I'm not too worried about the overhead since its a very simple hash lookup for and `Oid->int` mapping.)

Does that sound reasonable? Or, do you have any other ideas?

Reply:
sounds reasonable

Reply:
Actually, the `SELECT count(*) FROM reference_Table;` shouldn't really affect anything since it's outside of the transaction so it doesn't hold locks.

Reply:
That's true, but, it'd probably a little bit confusing. Because we're explicitly naming things with `RecordRelationAccess*`, and as a reader, the default expectation would probably be to see all the relation accesses. 

(Btw I've done some benchmarks locally, and I haven't observed any performance overhead. I think it's negligible. I'll do some proper benchmarks once we've all the pieces for fkey to reference tables)

(I cannot easily see any future use-cases for this relation accesses infrastructure, but, if we need to use it in the future, this could be confusing for sure).

That said, I have a slight preference to keep this as-is for now. Does that work for you?